language-plutus-core/exe/Main.hs

```diff
@@ -113,47 +179,115 @@ exampleSingle = ExampleSingle <$> exampleName
+parsePlcFile inp = do
+    contents <- getPlcInput inp
     let bsContents = (BSL.fromStrict . encodeUtf8 . T.pack) contents
+    case PLC.runQuoteT $ runExceptT (PLC.parseScoped bsContents) of
+        Left (errCheck :: PlcParserError) -> do
+            T.putStrLn $ PLC.prettyPlcDefText errCheck
+            exitFailure
+        Right (Left (errEval :: PlcParserError)) -> do
+            print errEval
+            exitFailure
+        Right (Right (p :: ParsedProgram)) ->
+            return p
+
+-- Read a PLC AST from a CBOR file
+getCborInput :: Input -> IO BSL.ByteString
+getCborInput StdInput         = BSL.getContents
+getCborInput (FileInput file) = BSL.readFile file
+
+-- Read either a PLC file or a CBOR file, depending on 'fmt'
+getProg :: Input -> Format -> IO ParsedProgram
+getProg inp fmt =
+    case fmt of
+      Plc  -> parsePlcFile inp
+      Cbor -> do
+               p <- getCborInput inp
+               return $ fakeAlexPosn <$ (deserialise p :: PlainProgram)
+                   where fakeAlexPosn = PLC.AlexPn 0 0 0
+
+
+---------------- Typechecking ----------------
+
+runTypecheck :: TypecheckOptions -> IO ()
+runTypecheck (TypecheckOptions inp fmt) = do
+    prog <- getProg inp fmt
     let cfg = PLC.defConfig
-    case (PLC.runQuoteT . PLC.parseTypecheck cfg) bsContents of
-        Left (e :: PLC.Error PLC.DefaultUni PLC.AlexPosn) -> do
+    case PLC.runQuoteT $ PLC.typecheckPipeline cfg prog of
+      Left (e :: PlcParserError) -> do
             T.putStrLn $ PLC.prettyPlcDefText e
             exitFailure
-        Right ty -> do
+      Right ty -> do
             T.putStrLn $ PLC.prettyPlcDefText ty
             exitSuccess
 
+
+---------------- Evaluation ----------------
+
 runEval :: EvalOptions -> IO ()
-runEval (EvalOptions inp mode) = do
-    contents <- getInput inp
-    let bsContents = (BSL.fromStrict . encodeUtf8 . T.pack) contents
-    let evalFn = case mode of
-            CK  -> first toException . PLC.extractEvaluationResult . PLC.evaluateCk
-            CEK -> first toException . PLC.extractEvaluationResult . PLC.evaluateCek mempty PLC.defaultCostModel
-    case evalFn . void . PLC.toTerm <$> PLC.runQuoteT (PLC.parseScoped bsContents) of
-        Left (errCheck :: PLC.Error PLC.DefaultUni PLC.AlexPosn) -> do
-            T.putStrLn $ PLC.prettyPlcDefText errCheck
-            exitFailure
-        Right (Left errEval) -> do
-            print errEval
-            exitFailure
-        Right (Right v) -> do
-            T.putStrLn $ PLC.prettyPlcDefText v
-            exitSuccess
+runEval (EvalOptions inp mode fmt) = do
+  prog <- getProg inp fmt
+  let evalFn = case mode of
+                 CK  -> first toException . PLC.extractEvaluationResult . PLC.evaluateCk
+                 CEK -> first toException . PLC.extractEvaluationResult . PLC.evaluateCek mempty PLC.defaultCostModel
+  case evalFn . void . PLC.toTerm $ prog of
+    Left errEval -> do
+      print errEval
+      exitFailure
```

Comment:
I wonder if we should use `unsafeEvaluateCk` here instead of carefully extracting an error only to print it and quit. We could throw it just as well. And `unsafeEvaluateCek`.

Reply:
I think that was just what it did before I started messing with it.  Also, does your final sentence terminate prematurely?

Reply:
> And `unsafeEvaluateCek`.

The suspense is unbearable.

Reply:
No, the diff suggests you didn't change the logic there.

Reply:
That does make it a bit shorter, and the behaviour is exactly the same.  The only difference is that with the version above we have the opportunity to intercept the error message and render it a bit more readably.  At the moment both versions give you parse errors like this:

```
ParseErrorE (Unexpected (TkName {tkLoc = AlexPn 43 2 29, tkName = "W", tkIdentifier = Unique {unUnique = 0}}))
```
That's not a big deal though, since I don't think many people will be using this.